plugins/parsers/bindata/parser.go

```diff
@@ -0,0 +1,216 @@
+package bindata
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+)
+
+const timeKey = "time"
+const timezone = "UTC"
+const defaultStringEncoding = "UTF-8"
+const defaultTimeFormat = "unix"
+
+// Field is a binary data field descriptor
+type Field struct {
+	Name string
+	Type string
+	Size uint
+}
+
+// BinData is a binary data parser
+type BinData struct {
+	metricName     string
+	timeFormat     string
+	endiannes      string
+	byteOrder      binary.ByteOrder
+	stringEncoding string
+	fields         []Field
+	DefaultTags    map[string]string
+}
+
+// Supported field types
+var fieldTypes = map[string]reflect.Type{
+	"bool":    reflect.TypeOf((*bool)(nil)).Elem(),
+	"uint8":   reflect.TypeOf((*uint8)(nil)).Elem(),
+	"int8":    reflect.TypeOf((*int8)(nil)).Elem(),
+	"uint16":  reflect.TypeOf((*uint16)(nil)).Elem(),
+	"int16":   reflect.TypeOf((*int16)(nil)).Elem(),
+	"uint32":  reflect.TypeOf((*uint32)(nil)).Elem(),
+	"int32":   reflect.TypeOf((*int32)(nil)).Elem(),
+	"uint64":  reflect.TypeOf((*uint64)(nil)).Elem(),
+	"int64":   reflect.TypeOf((*int64)(nil)).Elem(),
+	"float32": reflect.TypeOf((*float32)(nil)).Elem(),
+	"float64": reflect.TypeOf((*float64)(nil)).Elem(),
+	"string":  reflect.TypeOf((*string)(nil)).Elem(),
+	"padding": reflect.TypeOf((*[]byte)(nil)).Elem(),
+}
+
+// NewBinDataParser is BinData factory
+func NewBinDataParser(
+	metricName string,
+	timeFormat string,
+	endiannes string,
+	stringEncoding string,
+	fields []Field,
+	defaultTags map[string]string,
+) (*BinData, error) {
+
+	// Time format
+	switch timeFormat {
+	case "":
+		timeFormat = defaultTimeFormat
+	case "unix", "unix_ms", "unix_us", "unix_ns":
+	default:
+		return nil, fmt.Errorf("invalid time format %s", timeFormat)
+	}
+
+	// Endiannes
+	var byteOrder binary.ByteOrder
+	endiannes = strings.ToLower(endiannes)
+	switch endiannes {
+	case "", "be":
+		byteOrder = binary.BigEndian
+	case "le":
+		byteOrder = binary.LittleEndian
+	default:
+		return nil, fmt.Errorf("invalid bindata_endiannes %s", endiannes)
+	}
+
+	// String encoding
+	if stringEncoding == "" {
+		stringEncoding = defaultStringEncoding
+	}
+	stringEncoding = strings.ToUpper(stringEncoding)
+	if stringEncoding != defaultStringEncoding {
+		return nil, fmt.Errorf(`invalid string encoding %s`, stringEncoding)
+	}
+
+	// Field types, names and sizes
+	knownFields := make(map[string]bool)
+	for i := 0; i < len(fields); i++ {
+		fieldType, ok := fieldTypes[strings.ToLower(fields[i].Type)]
```

Comment:
```suggestion
	for i, field := range fields {
		fieldType, ok := fieldTypes[strings.ToLower(field.Type)]
```

Reply:
I think this away each `field` will be copied in a new var, I am fine with that - code reads better then, but at the end it will be slightly worse performance.